### PR TITLE
Fixes #18985: replace infinite scroll with pagination.

### PR DESCRIPTION
**Pull request by waldenraines**
http://projects.theforeman.org/issues/18985

### CODE REVIEW

**Comment by waldenraines** [comment]
Note that this requires https://github.com/theforeman/foreman/pull/4402 which was just merged this morning.

**Comment by waldenraines** [comment]
Screenshot:

screen shot 2017-03-23 at 10 16 31 am

**Comment by thomasmckay** [comment]
The "per page" setting should be included in the choice list and that choice selected when results are pulled.

For example, if I set entries_per_page to 10 then "10" should be added to the list of choices and "10" should be selected.

**Comment by thomasmckay** [comment]
The next page button works even if no further pages to visit (eg. I had 17 items w/ page size of 20 and clicked the next page ">" and it took me to a blank page.

The browser back button did not return me to page 1 results either.

**Comment by thomasmckay** [comment]
page size "All" gives per_page=NaN

**Comment by jlsherrill** [comment]
+1 to removing 'fixed headers'  it increases complexity and the need is much less with paged

**Comment by thomasmckay** [comment]
The root breadcrumb should keep the page state to return to (ie. if on page 4, root breadcrumb should return to page 4)

**Comment by thomasmckay** [comment]
+1 to removing fixed headers. The scrolling w/in scrolling is very awkward in paged mode.

**Comment by jlsherrill** [comment]
Thoughts on defaulting to the page size setting?

**Comment by waldenraines** [comment]
> The root breadcrumb should keep the page state to return to (ie. if on page 4, root breadcrumb should return to page 4)

That will be handled by http://projects.theforeman.org/issues/18485

**Comment by jlsherrill** [comment]
Selecting 'all' from the page selector doesn't work, ends up making a request for:
 "/katello/api/v2/packages?organization_id=1&page=3&paged=true&per_page=NaN&search="

NaNaNanNaNa

I vote we remove that as an option, its never going to be performant.

**Comment by waldenraines** [comment]
> I vote we remove that as an option, its never going to be performant.

I disagree because it gets more visibility into the slowness of our API such that we can hopefully get resources to improve it.  At the very least we ought to be doing server-side caching of API requests.

**Comment by jlsherrill** [comment]
we can't possibly write an api that responds to a request for 40,000 hosts at once in a timely manner.  Its simply not possible and will cause unintended DOS's against the satellite

**Comment by waldenraines** [comment]
> The browser back button did not return me to page 1 results either.

I have seen this problem elsewhere.  You get a blank page and this JS error?

```
Uncaught Error: [ng:btstrpd] App already bootstrapped with this element 'document'
http://errors.angularjs.org/1.5.5/ng/btstrpd?p0=document
    at angular.self.js?body=1:69
    at doBootstrap (angular.self.js?body=1:1728)
    at Object.bootstrap (angular.self.js?body=1:1768)
    at bastion-bootstrap.self.js?body=1:2
    at HTMLDocument.<anonymous> (jquery.turbolinks.self.js?body=1:30)
    at HTMLDocument.dispatch (bundle-ef1ff73….js:36)
    at HTMLDocument.g.handle (bundle-ef1ff73….js:35)
    at Object.trigger (bundle-ef1ff73….js:36)
    at HTMLDocument.<anonymous> (bundle-ef1ff73….js:36)
    at Function.each (bundle-ef1ff73….js:25)
```

See http://projects.theforeman.org/issues/19011

**Comment by waldenraines** [comment]
> Thoughts on defaulting to the page size setting?

What do you mean?  Doesn't it default to 20 for you?

**Comment by waldenraines** [comment]
> we can't possibly write an api that responds to a request for 40,000 hosts at once in a timely manner. Its simply not possible and will cause unintended DOS's against the server

That's fine, removing it makes my life easier anyway.  I wouldn't be surprised if we get requests for this though, it's pretty standard in most tables.

**Comment by waldenraines** [comment]
> Thoughts on defaulting to the page size setting?

Yeah, I can do that.  That's basically the same point that @thomasmckay brought up.

**Comment by waldenraines** [comment]
@jlsherrill @thomasmckay I believe that I have addressed all comments.

**Comment by ehelms** [comment]
Looks like this change is going to require a bump to the Bastion major version, @waldenraines would you agree? (Just thinking ahead of managing other updates).

**Comment by waldenraines** [comment]
> Looks like this change is going to require a bump to the Bastion major version, @waldenraines would you agree? (Just thinking ahead of managing other updates).

Yes, definitely, and hopefully the last bump to the Bastion major version as well.

**Comment by waldenraines** [comment]
Tests added, this is ready to merge pending review(s).

**Comment by waldenraines** [comment]
[test]

File: app/assets/javascripts/bastion/components/nutupane.factory.js [hunk 1]
```diff
@@ -44,7 +42,29 @@ angular.module('Bastion.components').factory('Nutupane',
...
+                if (params.paged) {
+                    $location.search("page", params.page).replace();
+                    $location.search("per_page", params['per_page']).replace();
+                }
+
+                if (params.search) {
+                    $location.search(self.searchKey, params.search).replace();
+                }
+
+                if (params['sort_by']) {
```
Comment (ehelms) [inline app/assets/javascripts/bastion/components/nutupane.factory.js]: Why use `[]` for access with this parameter when you used the dot operator for `params.search` ?

File: app/assets/javascripts/bastion/components/nutupane.factory.js [hunk 2]
```diff
@@ -320,27 +322,59 @@ angular.module('Bastion.components').factory('Nutupane',
...
-            self.table.nextPage = function () {
-                var table = self.table;
-                if (table.working || !table.hasMore()) {
-                    return false;
+            self.table.onFirstPage = function () {
+                return self.table.resource.page === 1;
+            };
+
+            self.table.onLastPage = function () {
+                return self.table.resource.page >= self.table.resource.subtotal / self.table.resource.per_page;
```
Comment (ehelms) [inline app/assets/javascripts/bastion/components/nutupane.factory.js]: Are we guaranteed these values will always be integers and not strings?

**Comment by ehelms** [comment]
@jlsherrill I'd argue we can write an APi that responds to 40k hosts, whether or not we can scale any given users infrastructure to handle that appropriately is the question. I know, semantics, and we should be optimizing for our minimum requirements. But there should be some expectation that if you are at certain scales, you should be scaling the app appropriately (assuming we supported ways to do that easily).

So does this mean that "Select All" should only select whats on the page for all pages?

**Comment by ehelms** [comment]
Ensure you have 40 of something (say products). Navigate to page 2 and then change the per page to greater than 40. Say 50 or 75. You end up with the table showing no results.

**Comment by ehelms** [comment]
Not a blocker, but it would be nice if it didn't refresh the entire table when switching pages and instead just spun loading over the set of rows. This would be less jarring and feel faster than everything disappearing and then re-appearing given I am just trying to sift through the pages of data.

**Comment by waldenraines** [comment]
> Ensure you have 40 of something (say products). Navigate to page 2 and then change the per page to greater than 40. Say 50 or 75. You end up with the table showing no results.

Do you mind checking to see what was in the API response in this case?  This may be an API bug.

File: app/assets/javascripts/bastion/components/nutupane.factory.js (hunk 1 above)
Comment (waldenraines) [inline app/assets/javascripts/bastion/components/nutupane.factory.js]: Because our linting used to fail on this, I guess it doesn't any longer.

**Comment by waldenraines** [comment]
> So does this mean that "Select All" should only select whats on the page for all pages?

No, I think we do the same select all we have done where you get the additional option to select everything (including stuff not on that particular page).

**Comment by waldenraines** [comment]
> Not a blocker, but it would be nice if it didn't refresh the entire table when switching pages and instead just spun loading over the set of rows. This would be less jarring and feel faster than everything disappearing and then re-appearing given I am just trying to sift through the pages of data.

See what you think about my latest update.  Instead of using the loading screen and hiding the table I just put the table mask on the table.  I know it's not exactly what you asked for but putting the spinner on the rows could prove more difficult as they are provided by each individual template in katello.  I could maybe do this with a directive in bastion but would like to revisit that as a follow up PR if that's okay with you.

**Comment by waldenraines** [comment]
[test]

**Comment by waldenraines** [comment]
[test]

**Comment by jlsherrill** [comment]
@ehelms given that we are on a postgresql database with roles and permission, and serializing data to json with the data that we send today, no i don't think we can reasonably do that, no matter how you setup the infrastructure.  Throw some or all of that out the window or adding caching at different layers, maybe you could get close, but I don't see it happening.

**Comment by waldenraines** [comment]
> @ehelms given that we are on a postgresql database with roles and permission, and serializing data to json with the data that we send today, no i don't think we can reasonably do that, no matter how you setup the infrastructure. Throw some or all of that out the window or adding caching at different layers, maybe you could get close, but I don't see it happening.

I thought the bottleneck was external API calls to pulp and candlepin?  

There are a lot of fast APIs that are backed by Postgres or other transactional databases.  I don't think the limiting factor is the DB unless it is the way in which our DB queries are constructed.  It has been my experience that using ORMs like active record often results in poorly written SQL with multiple joins on the same table etc.  I'm speaking as someone who has written [one such query with AR](https://github.com/Katello/katello/pull/6587) even though I know a fair amount about databases in general and PostgreSQL in particular.  

A related thought I just had: do we include vacuum statements as part of upgrades?  

> Also the UI rendering 40,000 (or 100,000) items would likely be very very sluggish

That's true.  This seems to be a case of whether or not we should allow the user to shoot themselves in the foot.  I almost always use "show all" when I can on various websites but I would be highly annoyed if the UI became unresponsive or difficult to use if I did that.  

We could allow "show all" conditionally based on the totals.  Thoughts?

**Comment by jlsherrill** [comment]
@waldenraines i'm fine with that, maybe allow all if its 200 or less or some sensible number like that

**Comment by ehelms** [comment]
@waldenraines I like the change very much! We could possibly have a spinner or something else but I think this is 90% of the way to make it feel as part of the page and fluid.

**Comment by ehelms** [comment]
Using this script, load in 100 products:

```
#!/usr/bin/env ruby

(3..100).each do |num|
  puts "Making #{num}"
  puts "curl -s -H 'Content-Type: application/json' -k -X POST -d '{\"organization_id\":1,\"name\":\"Product #{num}\"}' http://admin:changeme@0.0.0.0:3000/katello/api/v2/products"
  `curl -s -H 'Content-Type: application/json' -k -X POST -d '{\"organization_id\":1,\"name\":\"Product #{num}\"}' http://admin:changeme@0.0.0.0:3000/katello/api/v2/products`
end
```

 * Now, go to the products page
 * set the pgae size to 75
 * Click and go to the second page of results
 * Now in the search box type '79' and hit search

Note there are no results. See screenshot here where it thinks it is still on page 2 of 1.

screenshot from 2017-03-28 10-25-22

The API call being made for the search is:
```
http://192.168.121.8:3000/katello/api/v2/products?enabled=true&organization_id=1&page=2&paged=true&per_page=75&search=79&sort_by=name&sort_order=ASC
```

**Comment by waldenraines** [comment]
@ehelms @thomasmckay @jlsherrill I have addressed all issues.

**Comment by ehelms** [comment]
My issues have all been addressed. @thomasmckay @jlsherrill final ACks are y'alls

**Comment by thomasmckay** [comment]
ack

**Comment by jlsherrill** [comment]
ACK from me, one very minor issue:

you can type in a page # that doesn't exist and it will take you to a blank page.  i.e. with 10 items (all displayed), you can type in page 2 (even though there is only 1 page).

**Comment by jlsherrill** [comment]
and i'm okay merging and resolving that later(if it needs to be)

**Comment by waldenraines** [comment]
> you can type in a page # that doesn't exist and it will take you to a blank page. i.e. with 10 items (all displayed), you can type in page 2 (even though there is only 1 page).

Filed an issue for that, will fix as a followup to this:  http://projects.theforeman.org/issues/19071

**Pull request merged**